From 407a38aeb24be9fe76815c6d0dbb236b7a242d51 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Tue, 15 Jun 2010 02:35:06 +0100
Subject: [PATCH] usbhid: Reduce the race condition between disconnect and ioctl

hiddev_ioctl() always looks up the parent USB device, which may go
away even before exist is set to false.  Reduce the race condition by
only looking up the parent device when needed, i.e. for
HIDIOCGDEVINFO.  This should fix bug #511892.
---
 drivers/hid/usbhid/hiddev.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/hid/usbhid/hiddev.c b/drivers/hid/usbhid/hiddev.c
index 95cc192..26ee39b 100644
--- a/drivers/hid/usbhid/hiddev.c
+++ b/drivers/hid/usbhid/hiddev.c
@@ -545,7 +545,7 @@ static int hiddev_ioctl(struct inode *inode, struct file *file, unsigned int cmd
 	struct hiddev_list *list = file->private_data;
 	struct hiddev *hiddev = list->hiddev;
 	struct hid_device *hid = hiddev->hid;
-	struct usb_device *dev = hid_to_usb_dev(hid);
+	struct usb_device *dev;
 	struct hiddev_collection_info cinfo;
 	struct hiddev_report_info rinfo;
 	struct hiddev_field_info finfo;
@@ -579,6 +579,7 @@ static int hiddev_ioctl(struct inode *inode, struct file *file, unsigned int cmd
 		return hid->collection[i].usage;
 
 	case HIDIOCGDEVINFO:
+		dev = hid_to_usb_dev(hid);
 		dinfo.bustype = BUS_USB;
 		dinfo.busnum = dev->bus->busnum;
 		dinfo.devnum = dev->devnum;
-- 
1.7.1

